File: app/src/main/java/com/bitcoin/merchant/app/util/DateUtil.java
package com.bitcoin.merchant.app.util;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

public class DateUtil {
    private static DateUtil instance = null;
    private final SimpleDateFormat pastYearsFormat = new SimpleDateFormat("E dd MMM @ HH:mm");
    private final SimpleDateFormat currentYearFormat = new SimpleDateFormat("dd MMM yyyy");
    private final SimpleDateFormat previousDaysFormat = new SimpleDateFormat("E dd MMM @ HH:mm");
    private final SimpleDateFormat yesterdayFormat = new SimpleDateFormat("HH:mm");
    private final SimpleDateFormat todayFormat = new SimpleDateFormat("HH:mm");

    private DateUtil() {
    }

    public static DateUtil getInstance() {
        if (instance == null) {
            instance = new DateUtil();
        }
        return instance;
    }

    public String format(long timeInSec) {
        if (timeInSec == 0) {
            return "";
        }
        long hours24 = 60L * 60L * 24;
        long now = System.currentTimeMillis() / 1000L;
        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date(now * 1000L));
        int nowYear = cal.get(Calendar.YEAR);
        int nowDay = cal.get(Calendar.DAY_OF_MONTH);
        cal.setTime(new Date(timeInSec * 1000L));
        int thenYear = cal.get(Calendar.YEAR);
        int thenDay = cal.get(Calendar.DAY_OF_MONTH);
        // within 24h
        String ret;
        if (now - timeInSec < hours24) {
            if (thenDay < nowDay) {
                ret = "Yesterday @ " + yesterdayFormat.format(timeInSec * 1000L);
            } else {
                ret = "Today @ " + todayFormat.format(timeInSec * 1000L);
            }
        }
        // within 48h
        else if (now - timeInSec < (hours24 * 2)) {
            ret = previousDaysFormat.format(timeInSec * 1000L);
        } else {
            if (thenYear < nowYear) {
                ret = currentYearFormat.format(timeInSec * 1000L);
            } else {
                ret = pastYearsFormat.format(timeInSec * 1000L);
            }
        }
        return ret;
    }
}

File: app/src/main/java/com/bitcoin/merchant/app/util/DateUtil.kt
package com.bitcoin.merchant.app.util

import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

class DateUtil private constructor() {
    private val pastYearsFormat = SimpleDateFormat("E dd MMM @ HH:mm")
    private val currentYearFormat = SimpleDateFormat("dd MMM yyyy")
    private val previousDaysFormat = SimpleDateFormat("E dd MMM @ HH:mm")
    private val yesterdayFormat = SimpleDateFormat("HH:mm")
    private val todayFormat = SimpleDateFormat("HH:mm")
    private val hours24 = TimeUnit.HOURS.toMillis(24)
    fun format(timeInMillis: Long): String {
        if (timeInMillis == 0L) {
            return ""
        }
        val now = System.currentTimeMillis()
        val cal = Calendar.getInstance()
        cal.time = Date(now)
        val nowYear = cal[Calendar.YEAR]
        val nowDay = cal[Calendar.DAY_OF_MONTH]
        cal.time = Date(timeInMillis)
        return if (now - timeInMillis < hours24) {
            val thenDay = cal[Calendar.DAY_OF_MONTH]
            if (thenDay < nowDay) {
                "Yesterday @ " + yesterdayFormat.format(timeInMillis)
            } else {
                "Today @ " + todayFormat.format(timeInMillis)
            }
        } else if (now - timeInMillis < hours24 * 2) {
            previousDaysFormat.format(timeInMillis)
        } else {
            val thenYear = cal[Calendar.YEAR]
            if (thenYear < nowYear) {
                currentYearFormat.format(timeInMillis)
            } else {
                pastYearsFormat.format(timeInMillis)
            }
        }
    }

    fun formatHistorical(timeInMillis: Long): String {
        return pastYearsFormat.format(timeInMillis)
    }

    companion object {
        val instance: DateUtil by lazy { DateUtil() }
    }
}